### PR TITLE
Release maintenance cleanup

### PR DESCRIPTION
**Pull request by blink1073**
Simplifies the process of versioning and publishing by utilizing [bump2version](https://github.com/c4urself/bump2version) as well as `buildutils` scripts that add our own behavior on top if it.

Releases are now done with a single command that bumps the versions and publishes with the appropriate prerelease identifiers and tags.

Bumping the major versions of the JS packages is done with a single command, that could eventually be done by the author of a PR or using a bot.  The major version bump is smart enough to only bump if needed.

See the new release instructions for details.

## References
Fixes https://github.com/jupyterlab/jupyterlab/issues/5142.   Fixes https://github.com/jupyterlab/jupyterlab/issues/4368.  Related to https://github.com/jupyterlab/jupyterlab/issues/407.

## Code changes
Adds a script for handling version numbers and streamlines the patch release scripts.

## User-facing changes
None, these are developer-facing.

## Backwards-incompatible changes
None

### CODE REVIEW

**Comment by blink1073** [comment]
cc @saulshanabrook @jasongrout

**Comment by blink1073** [comment]
Was struggling with the syntax for using `commander`, now I see it is used throughout buildutils, ha.  Will come back to this later today.

**Comment by blink1073** [comment]
In the call yesterday we decided that the risk of information leak when publishing from CI was too great, so we're leaving it to the maintainer to run the publish command locally.

**Comment by blink1073** [comment]
This is done, just kicked a spurious failure on the Windows JS tests.

**Comment by blink1073** [comment]
Weird.  The Windows JS build passed on the second try but we still see the failing entry.

**Comment by blink1073** [comment]
Ping @jasongrout for review.

**Comment by jasongrout** [comment]
Thanks! I'm looking at this PR, and trying to understand some of the assumptions encoded in this PR:

1. We have a python version number and a number of js version numbers to manage.
2. A (python?) version number defined as `{major} {minor} {patch} {release} {build}`, where `{release}` is `alpha`, `candidate` or `final` and `{build}` is the pre-release number:
  * 1.2.3a4 has major 1, minor 2, patch 3, release `alpha` and build 4
  * 2.4.3 has major 1, minor 2, patch 3, release `final` and build 0
3. A python version number with release `final` will always have `build` 0.
4. Version bumps always go through the cycle `alpha` -> `candidate` -> `final` -> `alpha` -> `candidate` -> `final` -> `alpha` -> ...
5. js package numbers cannot be bumped when the python version has release equal to `final`
6. js `release` values track the python `release` values, e.g., bumping the major version of a js package when the python release is `alpha` bumps the js package to the next major number with alpha release.
7. If you try to bump a js package to the next major version, and it already has minor version 0 and a prerelease identifier, we assume it has already been bumped to a major version, so we don't bump it. This tries to make sure that once a js package has been bumped a major version in development, it won't be bumped again.

What is the workflow for some of the common scenarios we've seen?

A. A PR changes a single js package in a backwards-incompatible way, so its major version bump in the next release.
B. A PR changes a single js package such that it needs a minor bump in the next release.
C. A PR changes a single js package such that it needs a patch bump in the next release.

**Comment by blink1073** [comment]
Thanks for the feedback!  There are some subtle differences from what you said.  I also realized that if you bump a major version of a JS package, all of its dependencies need to be recursively bumped by a major version. I'll make it hat change and address your points when I pick this back up.

**Comment by blink1073** [comment]
1. We have a python version number and a number of js version numbers to manage.
2. A python version number defined as `{major} {minor} {patch} {release} {build}`, where `{release}` is `alpha`, `candidate` or `final` and `{build}` is the pre-release number:
    * 1.2.3a4 has major 1, minor 2, patch 3, release `alpha` and build 4
    * 2.4.3 has major 2, minor 4, patch 3, release `final` and build 0
3. A python version number with release `final` will always have `build` 0.
4. Version bumps always go through the cycle `alpha` -> `candidate` -> `final` -> `alpha` -> `candidate` -> `final` -> `alpha` -> ...
5. js package numbers cannot be bumped when the python version has release equal to `final`
6. js `release` values track the python `release` values, e.g., bumping the major version of a js package when the python release is `alpha` bumps the js package to the next *minor* number with alpha release.
7. If you try to bump a js package to the next major version, and it already has minor version 0 and a prerelease identifier, we assume it has already been bumped to a major version, so we don't bump it. This tries to make sure that once a js package has been bumped a major version in development, it won't be bumped again.
8. *TODO*  If you bump a js package by a major version, we also bump all of its dependencies by a major version, recursively.
9. *TODO* If you bump the python package by major and it is not in final, do not bump the js versions, but make them match the release value.

The workflow for some of the common scenarios we've seen:

A. A PR changes a single js package in a backwards-incompatible way, so its major version bump in the next release.
- We run `jlpm bump:js:major <package>` and continue developing.
B. A PR changes a single js package such that it needs a minor bump in the next release.
- All packages are assumed to have a minor bump with every release.  The only time the python package affects the js versions is when we switch from final -> alpha, at which time they all get bumped by a minor release.
C. A PR changes a single js package such that it needs a patch bump in the next release.
If this is a pre-release, there is no change.  If this is a patch to a previously released version, we run `jlpm patch:release <package>` on the `vxx.xx` branch and then run `jlpm publish:all`.

**Comment by blink1073** [comment]
This thread summarizes why we should bump all dependencies' major versions to be safe: https://github.com/semver/semver/issues/148#issuecomment-21970506.  In theory, we could only bump if we expose that dependency's api, but it can be more subtle than that because we require singletons for some packages, etc.

**Comment by jasongrout** [comment]
>     * All packages are assumed to have a minor bump with every release.  

seems to conflict with this:

> The only time the python package affects the js versions is when we switch from final -> alpha, at which time they all get bumped by a minor release.

Say we publish a final python release, then bump to alpha. You're saying the js packages all get a minor increment automatically. After we publish the first python alpha, do the js packages get another minor bump for the next python alpha?

**Comment by blink1073** [comment]
Python `final` -> `alpha 0` yields a JS minor bump to all packages (in alpha)
Python `alpha 0` -> alpha1 yields a build bump to all JS packages
Python `alpha 1` -> `release 0` makes all JS packages `release 0`
Python `release x` -> `final` makes all JS packages final
Python patch may or may not include patches to JS packages

**Comment by jasongrout** [comment]
> Python `final` -> `alpha 0` yields a minor bump to all JS packages (in `alpha`)

Do you mean the js packages go from x.y.z -> x.(y+1).0alpha0?

>  Python `alpha 1` -> `release 0` makes all JS packages switch to `rc` (e.g. `rc0`)
> Python `release x` -> `final` makes all JS packages final

In both of these, do you mean `candidate 0` and `candidate x`, rather than `release`?

**Comment by blink1073** [comment]
|  Action  | Command   | Result   | 
|---|---|---|
| Py x.y.z-> x.(y+1).a0 | jlpm bumpversion minor | JS packages a.b.c -> a.(b+1).c.alpha1 |
| Py x.y.z.a0-> x.y.z.a1 | jlpm bumpversion build | JS packages a.b.c.alpha.1 -> a.b.c.alpha2 |
| Py x.y.z.a1-> x.y.z.rc0 | jlpm bumpversion release | JS packages a.b.c.alpha.2 -> a.b.c.rc1 |
| Py x.y.z.rcx-> x.y.z | jlpm bumpversion release | JS packages a.b.c.rcx -> a.b.c
| Py x.y.z -> x.y.(z+1) | jlpm patch:release [...packages] | Selected JS packages a.b.c -> a.b.(c+1) |
| JS major release(s) | jlpm bump:js:major [...packages] | Affected JS packages bumped* |

* Python package is not affected, JS dependencies are also bumped a major version.  All affected packages
changed to match the current release type of the Python package (`alpha` or `rc`)

**Comment by jasongrout** [comment]
> All a.b.c -> a.(b+1).c.alpha1

should it go to a.(b+1).0.alpha0 ?

**Comment by blink1073** [comment]
> should it go to a.(b+1).0.alpha0 ? 

I don't actually know what Lerna does here :smile:

**Comment by blink1073** [comment]
Updated based on https://github.com/lerna/lerna/tree/master/commands/version

File: buildutils/src/utils.ts [hunk 1]
```diff
@@ -89,6 +89,65 @@ export function writeJSONFile(filePath: string, data: any): boolean {
...
+ * Pre-bump.
+ */
+export function prebump() {
+  // Ensure bump2version is installed (active fork of bumpversion)
+  run('python -m pip install bump2version');
+
+  // Make sure we start in a clean git state.
+  if (checkStatus('git diff --quiet') !== 0) {
+    throw new Error('Must be in a clean git state');
+  }
```
Comment (jasongrout) [inline buildutils/src/utils.ts]: Should we also run `git status` to make sure there are no untracked files?

File: RELEASE.md [hunk 2]
```diff
@@ -32,71 +32,29 @@ conda deactivate
 conda remove --all -y -n jlabrelease
 rm -rf jupyterlab
 
-conda create -c conda-forge -y -n jlabrelease notebook nodejs twine 'tornado<6'
+conda create -c conda-forge -y -n jlabrelease notebook nodejs
```
Comment (jasongrout) [inline RELEASE.md]: We should still install twine and we still insist on tornado<6, right?

**Comment by jasongrout** [comment]
I merged in master and resolved the conflicts.

**Comment by blink1073** [comment]
I think my table above should be in the release document, feel free to add that. 😁

**Comment by jasongrout** [comment]
This is still labeled WIP. I'm ready to merge it - was there anything else you wanted to do here?

**Comment by jasongrout** [comment]
> This is still labeled WIP. I'm ready to merge it - was there anything else you wanted to do here?

(besides adding that table to the docs, of course)

**Comment by blink1073** [comment]
Forgot, there are two TODOs above, but those can follow in a separate PR.

**Comment by blink1073** [comment]
I'm afk for the night

**Comment by jasongrout** [comment]
All right, tests pass, things look good. Let's give this a try!

**Pull request merged**